### ISSUE TITLE
Receiving 404 Error for links containing parenthesis. 

### ISSUE THREAD

**Issue by Roninnanaki** [cgiffard/node-simplecrawler#186]
When using queueURL() and providing a link containing '()' simple crawler responds with a 404.

**Comment by cgiffard** [comment]
This could be an issue with escaping. Would you mind posting the code you're using to queue the URL?

**Comment by Roninnanaki** [comment]
Sorry for the delayed response. I agree and think that it has to do something with how characters are being escaped. Here is an example of how I am using the crawler.

```
var crawler = new Crawler();

crawler.host = "localhost";

crawler.queueURL("http://localhost:8080/paren-(example).html");

crawler
    .on("fetch404", function (queueItem) {
        console.log(queueItem);
    })
    .on("fetcherror", function (queueItem) {
        console.log(queueItem);
    })
    .on("fetchcomplete", function (queueItem) {
        console.log(queueItem);
    })
    .on("complete", function () {
        done();
    })
    .start();
```

Below is the results of the console log. 

```
{ url: 'http://localhost:8080/paren-%28example%29.html',
  protocol: 'http',
  host: 'localhost',
  port: '8080',
  path: '/paren-%28example%29.html',
  depth: 2,
  fetched: true,
  status: 'notfound',
  stateData: 
   { requestLatency: 210,
     requestTime: 210,
     contentLength: 45495,
     contentType: 'text/html; charset=utf-8',
     code: 404,
     headers: 
      { 'content-security-policy-report-only': 'default-src https: \'unsafe-inline\' \'unsafe-eval\'',
        'x-content-security-policy-report-only': 'default-src https: \'unsafe-inline\' \'unsafe-eval\'',
        'x-frame-options': 'DENY',
        'x-xss-protection': '1; mode=block',
        'x-content-type-options': 'nosniff',
        'cache-control': 'no-cache, no-store, max-age=0, must-revalidate',
        'surrogate-control': 'no-cache, no-store, max-age=0, must-revalidate',
        'content-type': 'text/html; charset=utf-8',
        'content-length': '45495',
        etag: 'W/"b1b7-4p288R2BaF5nAThyGYwBzw"',
        date: 'Mon, 30 Nov 2015 20:53:21 GMT',
        connection: 'close' } },
  referrer: null }
```

Please let me know if you need anything else.

**Comment by cgiffard** [comment]
What server are you using? It seems like it doesn't understand the escaping — the URL appears correctly escaped for transport in the queue snippet you've posted.

**Comment by Roninnanaki** [comment]
We user express 4.X for our server along with an internal tool to help with handlebars. Sorry for the late response I have been trying to determine if it is some in either of those tools with no luck. 

I started reading about URLs and found that parens are an un-reserved character and do not nee to be escaped. [what-every-developer-should-know-about-urls](http://www.skorks.com/2010/05/what-every-developer-should-know-about-urls/) Which lead me to look at how you are escaping the character in the URL. I found that in processURL you have this try/catch. If I commented out normalize() then it leaves the parens alone, which fixes the problem. I realize that you may have the normalize there for another reason I don't yet understand, but I wonder if it were made optional or if you were to use [encodeURIComponet()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) if that wouldn't be just as good for simplecrawler.  
```
try {
        newURL =
            uri(URL)
                .absoluteTo(context.url);
                // .normalize();

        if (crawler.urlEncoding === "iso8859") {
            newURL = newURL.iso8859();
        }

    } catch (e) {
        // Couldn't process the URL, since URIjs choked on it.
        return false;
    }
```

**Comment by cgiffard** [comment]
Ah, good catch. Sounds like a bug with URIjs. I'll first try updating the dependency and seeing if they've fixed it — otherwise we should probably raise a bug over there.

**Comment by cgiffard** [comment]
It also sounds like there could be a parallel bug in node's URL parsing. I'm surprised I never saw this before. What version of node is it?

**Comment by Roninnanaki** [comment]
Right now we are on Node 0.12 but we are in the process of moving to 4.X.

**Comment by fredrikekelund** [comment]
[RFC 3986](http://tools.ietf.org/html/rfc3986) actually claims parens as reserved characters, and AFAICT that's the latest standard. This seems to be an error in the article you linked that [others have pointed](http://www.skorks.com/2010/05/what-every-developer-should-know-about-urls/#comment-587672379) out as well.

I actually did not find that removing the `normalize` call prevented the parens from being encoded. This was my test case:
```js
it("should not escape parens in URL's", function () {

    var links =
        discover("<a href='https://example.com/section(name)'>Link</a>");

    links.should.be.an("array");
    links.length.should.equal(1);
    links[0].should.equal("https://example.com/section(name)");
});
```

Maybe this is an error with your server somehow?

**Comment by fredrikekelund** [comment]
@Roninnanaki did you track down the cause of this problem? It doesn't seem to lie with URI.js (which is what simplecrawler relies on for the URL encoding)

**Comment by Roninnanaki** [comment]
@fredrikekelund I think that you maybe right about the server. Or it could be one of our other services that feed us the URLs. I did read that in the RFC 3986, my bad for not posting an update. Though at the time I was working on it once I comment the normalize() call our implementation of the crawler had no problem with the urls. At this point I no longer an on that project and won't be investigating any longer. As this was only an issue that I found in my teams situation I will close the issue. Thank you for all of your help and investigation.
